client/js/README.md

```diff
@@ -1,31 +1,170 @@
+
+```ts
+import { client } from "@gradio/client";
+
+const app = await client("user/space-name");
+const result = await app.predict("/predict");
+```
+
+##### `payload`
+
+The `payload` argument is generally optional but this depends on the API itself. If the API endpoint depends on values being passed in then it is required for the API request to succeed. The data that should be passed in is detailed on the "View API" page of a space.
+
+```ts
+import { client } from "@gradio/client";
+
+const app = await client("user/space-name");
+const result = await app.predict("/predict", [1, "Hello", "friends"]);
 ```
 
-**usage**
+<!-- Types for the different API inputs are provided and the example snippets on the "View API" will provide strong typing for the payload. These types need to be explicitly annotated as there currently no mechanism to automatically infer them due to the varied and dynamic nature of API endpoints. -->
+
+#### `submit`
+
+The `submit` method provides a more flexible way to call an API endpoint, providing you with status updates about the current progress of the prediction as well as supporting more complex endpoint types.
 
 ```ts
 import { client } from "@gradio/client";
 
-const app = client();
+const app = await client("user/space-name");
+const submission = app.submit("/predict", payload);
+```
+
+The `submit` method accepts the same [`endpoint`](#endpoint) and [`payload`](#payload) arguments as `predict`.
+
+The `submit` method does not return a promise and should not be awaited, instead it returns an object with a `on`, `off`, and `cancel` methods.
+
+##### `on`
+
+The `on` method allows you to subscribe to events related to the submitted API request. There are two types of event that can be subscribed to: `"data"` updates and `"status"` updates.
+
+`"data"` updates are issued when the API computes a value, the callback provided as the second argument will be called when such a value is sent to the client. The shape of the data depends on the way the API itself is constructed. This event may fire more than once if that endpoint supports emmitting new values over time.
 
-const prediction = app.predict(endpoint, payload);
+<!-- The type annotations provided by the "View API" page will ensure that the data payload is strongly typed. -->
 
-// listen for predictions
-prediction.on("data", (event: { data: Array<unknown>; type: "data" }) => {});
+`"status` updates are issued when the status of a request changes. This information allows you to offer feedback to users when the queue position of the request changes, or when the request changes from queued to processing.
 
-// listen for status updates
-prediction.on("status", (event: { data: Status; type: "data" }) => {});
+The status payload look like this:
 
+```ts
 interface Status {
+	queue: boolean;
```

Comment:
Sorry, we should have aligned on this sooner. 

These are the [status codes](https://github.com/gradio-app/gradio/blob/main/client/python/gradio_client/utils.py#L72) on the python client. We basically don't map any of the status messages to higher-level categories. Can we preserve that for js client?

This is the [shape](https://github.com/gradio-app/gradio/blob/main/client/python/gradio_client/utils.py#L144) of the status message in python. I think the only differences (js vs python) are

- status -> code
- position -> rank
- progress -> progress_data
- success: bool not present in js
- time not present in js

Can we align? Would be happy to make modifications to python client as well as it's pre 1.0 if there's consensus. cc @aliabid94 @abidlabs

Reply:
- Add code - raw API code
- rename status to stage
- add time (timestamp)
- rename progress to progress_data
- add success (from api)

Reply:
- Add code - raw API code
- rename status to stage
- add time (timestamp)
- rename progress to progress_data
- add success (from api)